### ISSUE TITLE
BatchNorm does not correctly compute mean/variance across workers when they have different batch sizes

### ISSUE THREAD

**Issue by ppwwyyxx** [google/flax#2417]
### System information
Not relevant

### Problem you have encountered:
The code at 
https://github.com/google/flax/blob/9fe386508e35c31ca3e1f151bf1a4a87d71486ae/flax/linen/normalization.py#L82-L93
shows that the cross-device mean is computed by the "mean of per-device mean". This is correct only if each device has the same number of elements to normalize over. If they have different batch sizes (or different spatial dimensions in case of images), the result mean & variance is no longer the same as normalizing all samples together.

### What you expected to happen:

Although this feels mathematically wrong, in my experience in detection models, this behavior is __OK__  as long as we compute correct gradients. I also mentioned this in Appendix A.6 of https://arxiv.org/pdf/2105.07576.pdf.

So to me it may be fine to keep things as is, as long as this is documented clearly. But fixing the behavior is probably better.

**Comment by cgarciae** [comment]
Hey @ppwwyyxx

>  If they have different batch sizes (or different spatial dimensions in case of images), the result mean & variance is no longer the same as normalizing all samples together.

To my knowledge there is no way of getting different array shapes per device in jax given the SPMD design. Is there a realistic scenario you have in mind?

**Comment by ppwwyyxx** [comment]
@cgarciae I'm new to jax so I might be wrong, but to my knowledge, when using `pmap` in a multi-host / multi-process set up, each host independently processes its own data and can have data with different shapes, e.g.:
```
      x = jnp.ones((2, 4))

      def func(x):
          y = x[:jax.process_index()+1]
          print("Y:", y.shape, y)
          sum = jax.lax.psum(y.sum(), axis_name="i")
          return sum

      print(jax.pmap(func, axis_name="i")(x))
```

**Comment by jheek** [comment]
> when using pmap in a multi-host / multi-process environment, each host independently processes its own data and can have data with different shapes

Each host compiles the program independently indeed. But correct behavior depends on each host compiling an identical program (same ops and shapes). It might sometimes work if you violate this assumption but it is by not garantueed.

**Comment by ppwwyyxx** [comment]
Fair enough, I'll close this then!
